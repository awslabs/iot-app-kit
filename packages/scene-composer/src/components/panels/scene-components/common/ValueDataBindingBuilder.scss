.tm-copy-select-flex {
    display: flex;
    flex-wrap: nowrap;
}

.tm-full-width {
    width: 100%;
}