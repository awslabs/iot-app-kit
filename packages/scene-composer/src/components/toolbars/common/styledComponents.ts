import styled, { css } from 'styled-components';
import { Box, Icon as PolarisIcon } from '@awsui/components-react';
import {
  colorBackgroundDropdownItemDefault,
  colorBackgroundDropdownItemHover,
  colorBackgroundItemSelected,
  colorBorderDividerDefault,
  colorTextHeadingDefault,
} from '@awsui/design-tokens';

import { ToolbarItemIconProps, ToolbarItemOptions } from './types';

const DEFAULT_TEXT_PADDING = 20;
const ITEM_DIVIDER_WIDTH = 1;
const GROUP_DIVIDER_WIDTH = 3;

export const CornerAdornment = styled.div`
  position: absolute;
  bottom: 0;
  right: 0;
  width: 16%;
  height: 16%;
  pointer-events: none;
  z-index: 1;
  color: ${colorTextHeadingDefault};
  > svg {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    transform: rotate(90deg);
  }
`;

export const Icon = styled(PolarisIcon)<ToolbarItemIconProps>`
  shape-rendering: geometricPrecision;
  transform: ${({ scale = 1 }) => css`scale(${scale});`};
  transform: ${({ scale = 1, isMirrored }) => isMirrored && css`scaleX(${-scale});`};
`;

export const ToolbarItemContainer = styled.div<Pick<ToolbarItemOptions, 'isDisabled' | 'isSelected'>>`
  flex: none;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  min-width: 40px;
  height: 40px;
  text-decoration: none;
  cursor: ${({ isDisabled, isSelected }) => (isDisabled || isSelected ? 'default' : 'pointer')};
  pointer-events: ${({ isDisabled, isSelected }) => (isDisabled || isSelected ? 'none' : 'initial')};
  background-color: ${({ isSelected }) =>
    isSelected ? colorBackgroundItemSelected : colorBackgroundDropdownItemDefault};
  border-top: ${ITEM_DIVIDER_WIDTH}px solid ${colorBorderDividerDefault};

  &:hover {
    background-color: ${({ isDisabled, isSelected }) => !isDisabled && !isSelected && colorBackgroundDropdownItemHover};
  }

  &:first-of-type {
    border-top: 0;
  }
`;

export const ToolbarItemGroup = styled.div`
  position: relative;
  display: flex;
  flex-direction: column;
  border-top: ${GROUP_DIVIDER_WIDTH}px solid ${colorBorderDividerDefault};

  &:first-of-type {
    border-top: 0;
  }
`;

export const ToolbarItemIcon = styled.div<ToolbarItemIconProps>`
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 40px;
  height: 40px;
`;

export const ToolbarItemMenu = styled.div<{
  isOpen?: boolean;
  orientation?: 'horizontal' | 'vertical';
}>`
  display: none;
  position: absolute;
  left: 100%;
  top: 0;
  flex-direction: ${({ orientation }) => (orientation === 'horizontal' ? 'row' : 'column')};
  background-color: ${colorBackgroundDropdownItemDefault};
  box-shadow: ${({ theme }) => theme.boxShadow};
  user-select: none;
  z-index: 1;

  ${({ orientation }) => {
    if (orientation === 'horizontal') {
      return `
        & > div {
          border-left: 1px solid ${colorBorderDividerDefault};
          &:first-of-type {
            border-left: 0;
          }
        }
      `;
    } else {
      return `
        & > div {
          border-top: 1px solid ${colorBorderDividerDefault};
          &:first-of-type {
            border-top: 0;
          }
        }
      `;
    }
  }}

  ${({ isOpen }) => {
    return isOpen ? `display: flex; cursor: default;` : '';
  }}
`;

export const ToolbarItemText = styled(Box)<{ leftPadding?: number; rightPadding?: number }>`
  padding: 0 ${({ rightPadding = DEFAULT_TEXT_PADDING }) => rightPadding}px 0
    ${({ leftPadding = DEFAULT_TEXT_PADDING }) => leftPadding}px;
  white-space: nowrap;
`;

export const SubMenuIconContainer = styled.div`
  display: flex;
  flex: 1;
  flex-direction: row;
  justify-content: flex-end;
  padding-right: 6px;
`;
