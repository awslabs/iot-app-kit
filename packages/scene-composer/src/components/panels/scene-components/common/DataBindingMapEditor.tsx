import { Box, Button, SpaceBetween } from '@awsui/components-react';
import { isEmpty } from 'lodash';
import React, { useCallback } from 'react';
import { useIntl } from 'react-intl';
import styled from 'styled-components';

import { IValueDataBinding, IValueDataBindingProvider } from '../../../../interfaces';
import { Component } from '../../../../models/SceneModels';
import { ISceneComponentInternal } from '../../../../store';
import { Divider } from '../../../Divider';
import { DataBindingMapNameEditor } from '../data-overlay/DataBindingMapNameEditor';

import { ValueDataBindingBuilder } from './ValueDataBindingBuilder';

export interface ComponentWithDataBindings extends ISceneComponentInternal {
  valueDataBindings: Component.IDataBindingMap[] | Component.ValueDataBindingNamedMap[];
}

interface IDataBindingMapEditorProps {
  hasBindingName: boolean;
  valueDataBindingProvider: IValueDataBindingProvider | undefined;
  component: ComponentWithDataBindings;
  onUpdateCallback: (componentPartial: Partial<ComponentWithDataBindings>, replace?: boolean | undefined) => void;
  numFields?: number;
  hasRemoveButton?: boolean;
  skipFirstDivider?: boolean;
  allowPartialBinding?: boolean;
}

const RemoveButtonContainer = styled.div`
  float: right;
  height: 12px;
`;
const Spacing = styled.div`
  height: 8px;
`;

export const DataBindingMapEditor: React.FC<IDataBindingMapEditorProps> = ({
  hasBindingName,
  valueDataBindingProvider,
  component,
  numFields,
  hasRemoveButton,
  onUpdateCallback,
  skipFirstDivider,
  allowPartialBinding,
}) => {
  const intl = useIntl();

  const onBindingChange = useCallback(
    (valueDataBinding: IValueDataBinding, index) => {
      // we don't want to merge the dataBindingContext, so we'll need to manually replace it
      const updatedComponent = {
        ...component,
        valueDataBindings: [...component.valueDataBindings],
      };
      updatedComponent.valueDataBindings[index] = {
        ...updatedComponent.valueDataBindings[index],
        valueDataBinding,
      };
      onUpdateCallback(updatedComponent, true);
    },
    [component.valueDataBindings, onUpdateCallback],
  );

  const onRemoveBinding = useCallback(
    (index) => {
      const newBindings = component.valueDataBindings.filter((_, i) => i !== index);
      onUpdateCallback({ ...component, valueDataBindings: newBindings }, true);
    },
    [component.valueDataBindings, onUpdateCallback],
  );

  return (
    <SpaceBetween size='s'>
      {valueDataBindingProvider && (
        <>
          {!isEmpty(component.valueDataBindings) &&
            component.valueDataBindings.map((binding, index) => (
              <Box key={index}>
                {(index > 0 || !skipFirstDivider) && <Divider />}

                {hasRemoveButton && (
                  <RemoveButtonContainer>
                    <Button
                      ariaLabel={intl.formatMessage({
                        defaultMessage: 'Remove data binding',
                        description: 'Button label',
                      })}
                      data-testid='remove-binding-button'
                      iconName='close'
                      variant='icon'
                      iconAlign='right'
                      onClick={() => onRemoveBinding(index)}
                    />
                  </RemoveButtonContainer>
                )}
                <Spacing />
                {hasBindingName && (
                  <DataBindingMapNameEditor
                    bindingName={(binding as Component.ValueDataBindingNamedMap).bindingName}
                    index={index}
                    valueDataBindings={component.valueDataBindings as Component.ValueDataBindingNamedMap[]}
                    onUpdateCallback={onUpdateCallback}
                  />
                )}

                <Box margin={{ vertical: 's' }}>
                  <ValueDataBindingBuilder
                    allowPartialBinding={allowPartialBinding}
                    componentRef={component.ref}
                    binding={binding.valueDataBinding}
                    numFields={numFields}
                    valueDataBindingProvider={valueDataBindingProvider}
                    onChange={(v) => onBindingChange(v, index)}
                  />
                </Box>
              </Box>
            ))}
        </>
      )}
    </SpaceBetween>
  );
};
