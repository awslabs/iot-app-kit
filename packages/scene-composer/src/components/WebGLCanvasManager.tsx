import * as THREE from 'three';
import * as awsui from '@cloudscape-design/design-tokens';
import React, { useContext, useEffect, useRef } from 'react';
import { GizmoHelper, GizmoViewport } from '@react-three/drei';
import { ThreeEvent, useThree } from '@react-three/fiber';
import { MatterportModel } from '@matterport/r3f/dist';

import { KnownSceneProperty, COMPOSER_FEATURES } from '../interfaces';
import useLifecycleLogging from '../logger/react-logger/hooks/useLifecycleLogging';
import { useEditorState, useSceneDocument, accessStore } from '../store';
import { sceneComposerIdContext } from '../common/sceneComposerIdContext';
import { hexColorFromDesignToken } from '../utils/styleUtils';
import { Layers, ROOT_OBJECT_3D_NAME, MAX_CLICK_DISTANCE } from '../common/constants';
import { getGlobalSettings } from '../common/GlobalSettings';
import { ViewCursorWidget } from '../augmentations/components/three-fiber/viewpoint/ViewCursorWidget';
import { getIntersectionTransform } from '../utils/raycastUtils';
import { createNodeWithPositionAndNormal } from '../utils/nodeUtils';
import { EnvironmentLoadingManager } from '../common/loadingManagers';
import useMatterportViewer from '../hooks/useMatterportViewer';

import Environment, { presets } from './three-fiber/Environment';
import Fog from './three-fiber/Fog';
import GroundPlane from './three-fiber/GroundPlane';
import { StatsWindow } from './three-fiber/StatsWindow';
import EntityGroup from './three-fiber/EntityGroup';
import { EditorMainCamera } from './three-fiber/EditorCamera';
import { EditorTransformControls } from './three-fiber/EditorTransformControls';
import { SceneInfoView } from './three-fiber/SceneInfoView';
import SceneBackground from './three-fiber/SceneBackground';
import IntlProvider from './IntlProvider';

const GIZMO_MARGIN: [number, number] = [72, 72];

const envLoaderExtension = (loader: THREE.Loader) => {
  loader.manager = EnvironmentLoadingManager;
};

export const WebGLCanvasManager: React.FC = () => {
  const log = useLifecycleLogging('WebGLCanvasManager');

  const sceneComposerId = useContext(sceneComposerIdContext);
  const { isEditing, addingWidget, setAddingWidget } = useEditorState(sceneComposerId);
  const sceneAppearanceEnabled = getGlobalSettings().featureConfig[COMPOSER_FEATURES.SceneAppearance];
  const { enableMatterportViewer } = useMatterportViewer();
  const { document, getSceneNodeByRef, getSceneProperty } = useSceneDocument(sceneComposerId);
  const appendSceneNode = accessStore(sceneComposerId)((state) => state.appendSceneNode);
  const { gl } = useThree();

  const domRef = useRef<HTMLElement>(gl.domElement.parentElement);
  const environmentPreset = getSceneProperty<string>(KnownSceneProperty.EnvironmentPreset);
  const rootNodeRefs = document.rootNodeRefs;

  const gridHelperRef = useRef<THREE.GridHelper>(null);

  useEffect(() => {
    if (!!environmentPreset && !(environmentPreset in presets)) {
      log?.error('Environment preset must be one of: ' + Object.keys(presets).join(', '));
    }
  }, [environmentPreset]);

  const onClick = (e: ThreeEvent<MouseEvent>) => {
    if (e.delta <= MAX_CLICK_DISTANCE && addingWidget && e.intersections.length > 0) {
      const { position } = getIntersectionTransform(e.intersections[0]);
      const newWidgetNode = createNodeWithPositionAndNormal(addingWidget, position, new THREE.Vector3());

      setAddingWidget(undefined);
      appendSceneNode(newWidgetNode);

      e.stopPropagation();
    }
  };

  useEffect(() => {
    const gridHelper = gridHelperRef.current;
    if (gridHelper) {
      gridHelper.layers.disable(Layers.RaycastAndRender);
      gridHelper.layers.enable(Layers.RenderOnly);
    }
  }, [gridHelperRef.current]);

  return (
    <React.Fragment>
      {sceneAppearanceEnabled && (
        <React.Fragment>
          <Fog />
          <SceneBackground />
        </React.Fragment>
      )}
      <EditorMainCamera />
      {environmentPreset && environmentPreset in presets && (
        <Environment preset={environmentPreset} extensions={envLoaderExtension} />
      )}
      <group name={ROOT_OBJECT_3D_NAME} dispose={null}>
        {rootNodeRefs &&
          rootNodeRefs.map((rootNodeRef) => {
            const node = getSceneNodeByRef(rootNodeRef);
            return node && <EntityGroup key={rootNodeRef} node={node} />;
          })}
      </group>
      <GroundPlane />
      {isEditing() && (
        <React.Fragment>
          <EditorTransformControls />
          <GizmoHelper alignment='bottom-right' margin={GIZMO_MARGIN} renderPriority={0}>
            <GizmoViewport
              axisColors={[
                hexColorFromDesignToken(awsui.colorBackgroundNotificationRed),
                hexColorFromDesignToken(awsui.colorBackgroundNotificationGreen),
                hexColorFromDesignToken(awsui.colorBackgroundNotificationBlue),
              ]}
            />
          </GizmoHelper>
          <ViewCursorWidget />
          <React.Fragment>
            <gridHelper
              ref={gridHelperRef}
              position={new THREE.Vector3(0, 0.001, 0)}
              args={[
                1000 /* size */,
                500 /* grid# */,
                hexColorFromDesignToken(awsui.colorTextInputPlaceholder) /* center line color */,
                hexColorFromDesignToken(awsui.colorTextInputDisabled) /* grid color */,
              ]}
            />
            {enableMatterportViewer && <MatterportModel onClick={onClick} />}
          </React.Fragment>
          <IntlProvider locale={getGlobalSettings().locale}>
            <SceneInfoView />
          </IntlProvider>

          {getGlobalSettings().debugMode && <StatsWindow parent={domRef} />}
        </React.Fragment>
      )}
    </React.Fragment>
  );
};
