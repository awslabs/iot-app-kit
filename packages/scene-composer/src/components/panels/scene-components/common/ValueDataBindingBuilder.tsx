import { Autosuggest, Box, FormField, Select, SpaceBetween } from '@awsui/components-react';
import React, { useContext, useEffect, useState } from 'react';
import { defineMessages, useIntl } from 'react-intl';

import { EMPTY_VALUE_DATA_BINDING_PROVIDER_STATE } from '../../../../common/constants';
import { sceneComposerIdContext } from '../../../../common/sceneComposerIdContext';
import {
  IDataFieldOption,
  IValueDataBinding,
  IValueDataBindingProvider,
  IValueDataBindingProviderState,
} from '../../../../interfaces';
import useLifecycleLogging from '../../../../logger/react-logger/hooks/useLifecycleLogging';
import { useStore } from '../../../../store';
import { dataBindingConfigSelector } from '../../../../utils/dataBindingTemplateUtils';
import { pascalCase } from '../../../../utils/stringUtils';

export const ENTITY_ID_INDEX = 0;
export const COMPONENT_NAME_INDEX = 1;
export const PROPERTY_NAME_INDEX = 2;
export interface IValueDataBindingBuilderProps {
  componentRef: string;
  binding?: IValueDataBinding;
  allowPartialBinding?: boolean;
  numFields?: number;
  valueDataBindingProvider: IValueDataBindingProvider;
  onChange: (valueDataBinding: IValueDataBinding) => void;
}

export const ValueDataBindingBuilder: React.FC<IValueDataBindingBuilderProps> = ({
  componentRef,
  binding,
  allowPartialBinding,
  numFields,
  valueDataBindingProvider,
  onChange,
}: IValueDataBindingBuilderProps) => {
  useLifecycleLogging(`ValueDataBindingBuilder[${componentRef}]`);

  const sceneComposerId = useContext(sceneComposerIdContext);
  const dataBindingConfig = useStore(sceneComposerId)(dataBindingConfigSelector);
  const valueDataBindingStore = valueDataBindingProvider.useStore(false);
  const [builderState, setBuilderState] = useState<IValueDataBindingProviderState>(
    EMPTY_VALUE_DATA_BINDING_PROVIDER_STATE,
  );
  const [autoSuggestValue, setAutoSuggestValue] = useState('');

  const intl = useIntl();
  const i18nDataBindingLabelKeysStrings = defineMessages({
    entityId: {
      defaultMessage: 'Entity Id',
      description: 'Form Field label',
    },
    componentName: {
      defaultMessage: 'Component Name',
      description: 'Form Field label',
    },
    propertyName: {
      defaultMessage: 'Property Name',
      description: 'Form Field label',
    },
    entityPath: {
      defaultMessage: 'Entity Path',
      description: 'Form Field label',
    },
    componentTypeId: {
      defaultMessage: 'Component Type Id',
      description: 'Form Field label',
    },
  });

  const filterBuilderState = (state: IValueDataBindingProviderState) => {
    if (numFields) {
      return {
        definitions: state.definitions.slice(0, numFields),
        selectedOptions: state.selectedOptions.slice(0, numFields),
      };
    }
    return state;
  };

  useEffect(() => {
    // Subscribe to the changes
    valueDataBindingStore.setOnStateChangedListener((state) => {
      setBuilderState(filterBuilderState(state));
    });
    return () => valueDataBindingStore.setOnStateChangedListener(undefined);
  }, [valueDataBindingProvider]);

  useEffect(() => {
    // Initiate the provider
    const state = valueDataBindingStore.setBinding(componentRef, binding, dataBindingConfig);
    setBuilderState(filterBuilderState(state));
    setAutoSuggestValue(state.selectedOptions[ENTITY_ID_INDEX]?.value || '');
  }, [componentRef, binding, valueDataBindingProvider, dataBindingConfig]);

  return (
    <React.Fragment>
      <SpaceBetween size='s'>
        {builderState.definitions.map((definition, index) => {
          const { options, state } = definition;
          let selectedOption: IDataFieldOption | null = null;
          if (index < builderState.selectedOptions.length) {
            selectedOption = builderState.selectedOptions[index];
          }
          const disabled = state !== 'ready';
          if (definition.fieldName === 'entityId') {
            return (
              <FormField
                label={
                  intl.formatMessage(i18nDataBindingLabelKeysStrings[definition.fieldName]) ||
                  pascalCase(definition.fieldName)
                }
                key={definition.fieldName}
              >
                <Autosuggest
                  data-testid='select-entityId'
                  options={options}
                  enteredTextLabel={(item) => `${item}`}
                  virtualScroll
                  value={autoSuggestValue}
                  invalid={builderState.errors?.invalidEntityId}
                  onChange={async (event) => {
                    setAutoSuggestValue(event.detail.value);
                    await valueDataBindingStore.updateSelection(
                      definition.fieldName,
                      { value: event.detail.value },
                      dataBindingConfig,
                    );
                    if (allowPartialBinding) {
                      // take a snapshot of the callback in case the onChange function changed during
                      // the async processing.
                      const memorizedOnChange = onChange;
                      valueDataBindingStore.createBinding().then((value) => {
                        if (value) {
                          memorizedOnChange(value);
                        }
                      });
                    }
                  }}
                />
              </FormField>
            );
          }
          return (
            <FormField
              label={
                <Box>
                  {intl.formatMessage(i18nDataBindingLabelKeysStrings[definition.fieldName]) ||
                    pascalCase(definition.fieldName)}
                  {allowPartialBinding && (
                    <Box display='inline'>
                      <i>{` - ${intl.formatMessage({
                        defaultMessage: 'optional',
                        description: 'FormField label suffix',
                      })}`}</i>
                    </Box>
                  )}
                </Box>
              }
              key={definition.fieldName}
            >
              <Select
                data-testid='value-data-binding-builder-select'
                selectedOption={selectedOption}
                onChange={async (e) => {
                  await valueDataBindingStore.updateSelection(
                    definition.fieldName,
                    e.detail.selectedOption,
                    dataBindingConfig,
                  );
                  // trigger data binding context creation if the selected option is in the last slot or partial binding
                  // is allowed
                  if (index === builderState.definitions.length - 1 || allowPartialBinding) {
                    // take a snapshot of the callback in case the onChange function changed during
                    // the async processing.
                    const memorizedOnChange = onChange;
                    valueDataBindingStore.createBinding().then((value) => {
                      if (value) {
                        memorizedOnChange(value);
                      }
                    });
                  }
                }}
                options={options}
                disabled={disabled}
                placeholder={intl.formatMessage({ defaultMessage: 'Select an option', description: 'placeholder' })}
              />
            </FormField>
          );
        })}
      </SpaceBetween>
    </React.Fragment>
  );
};
