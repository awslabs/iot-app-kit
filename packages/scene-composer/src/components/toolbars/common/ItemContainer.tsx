import { isEmpty } from 'lodash';
import React, { ReactNode, useCallback, useState } from 'react';

import { getGlobalSettings } from '../../../common/GlobalSettings';

import {
  Icon,
  SubMenuIconContainer,
  ToolbarItemContainer,
  ToolbarItemIcon,
  ToolbarItemMenu,
  ToolbarItemText,
} from './styledComponents';
import { ToolbarItemOptions, ToolbarItemOrientation, ToolbarItemType } from './types';

interface ToolbarItemContainerProps {
  item: ToolbarItemOptions;
  type: ToolbarItemType;
  children?: ReactNode;
  onPointerDown?: React.PointerEventHandler<HTMLDivElement>;
  onPointerEnter?: React.PointerEventHandler<HTMLDivElement>;
  onPointerLeave?: React.PointerEventHandler<HTMLDivElement>;
  onPointerUp?: React.PointerEventHandler<HTMLDivElement>;
  orientation?: ToolbarItemOrientation;
  onItemClick?: (item: ToolbarItemOptions) => void;
  disableSelectedStyle?: boolean;
}

// Solves forwardRef with props and eslint issue. Explictly setting the props type (twice) silences eslint
// error "'XXX' is missing in props validation." See https://stackoverflow.com/a/66948051
export const ItemContainer = React.forwardRef<HTMLDivElement, ToolbarItemContainerProps>(
  (
    {
      children,
      item,
      onPointerDown,
      onPointerEnter,
      onPointerLeave,
      onPointerUp,
      orientation,
      onItemClick,
      type,
      disableSelectedStyle,
    }: ToolbarItemContainerProps,
    ref,
  ) => {
    const [hoveredItemId, setHoveredItemId] = useState<string>('');

    const pointerDown = useCallback(
      (e) => {
        if (e.pointerType !== 'mouse' && type !== 'button') {
          setHoveredItemId(item.uuid);
        }
        onPointerDown?.(e);
      },
      [type, item.uuid, onPointerDown],
    );

    const pointerUp = useCallback(
      (e) => {
        onPointerUp?.(e);
        onItemClick?.(item);
        e.stopPropagation();
      },
      [item, onPointerUp],
    );

    const pointerEnter = useCallback(
      (e) => {
        if (e.pointerType === 'mouse' && type !== 'button') {
          setHoveredItemId(item.uuid);
        }
        onPointerEnter?.(e);
      },
      [type, item.uuid, onPointerEnter],
    );

    const pointerLeave = useCallback(
      (e) => {
        if (e.pointerType === 'mouse' && type !== 'button') {
          setHoveredItemId('');
        }
        onPointerLeave?.(e);
      },
      [type, onPointerLeave],
    );

    if (item.feature) {
      const featureEnabled = getGlobalSettings().featureConfig[item.feature.name];
      if (!featureEnabled) {
        return null;
      }
    }

    return (
      <ToolbarItemContainer
        isDisabled={item.isDisabled}
        isSelected={!disableSelectedStyle && item.isSelected}
        onPointerDown={pointerDown}
        onPointerEnter={pointerEnter}
        onPointerLeave={pointerLeave}
        onPointerUp={pointerUp}
        data-testid={item.uuid}
        ref={ref}
      >
        {item.icon && (
          <ToolbarItemIcon>
            <span role='img' aria-label={item.label} title={item.label}>
              <Icon {...item.icon} variant={item.isDisabled ? 'disabled' : 'normal'} />
            </span>
          </ToolbarItemIcon>
        )}
        {item.text && (
          <ToolbarItemText
            color={item.isDisabled ? 'text-status-inactive' : 'inherit'}
            leftPadding={item.icon ? 0 : undefined}
            variant='small'
          >
            {item.text}
          </ToolbarItemText>
        )}
        {!isEmpty(item.subItems) && (
          <SubMenuIconContainer>
            <Icon name='caret-right-filled' />
          </SubMenuIconContainer>
        )}
        {!isEmpty(item.subItems) && (
          <ToolbarItemMenu isOpen={hoveredItemId === item.uuid} orientation={orientation}>
            {item.subItems?.map((subItem) => {
              return (
                <ItemContainer
                  key={subItem.uuid}
                  onPointerDown={onPointerDown}
                  onPointerEnter={onPointerEnter}
                  onPointerLeave={onPointerLeave}
                  onPointerUp={onPointerUp}
                  onItemClick={onItemClick}
                  item={subItem}
                  type={type}
                />
              );
            })}
          </ToolbarItemMenu>
        )}

        {children}
      </ToolbarItemContainer>
    );
  },
);
