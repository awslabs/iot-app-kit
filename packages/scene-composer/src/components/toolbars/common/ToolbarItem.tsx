import React, { Fragment, useEffect, useMemo, useRef, useState } from 'react';

import { CornerTriangleSvg } from '../../../assets/svgs';

import { ItemContainer } from './ItemContainer';
import { CornerAdornment, ToolbarItemMenu } from './styledComponents';
import { ToolbarItemOptions, ToolbarItemOrientation, ToolbarItemType } from './types';

interface ToolbarItemProps<T extends ToolbarItemOptions> {
  items: T[];
  type: ToolbarItemType;
  initialSelectedItem?: T;
  onClick?: (item: T) => void;
  orientation?: ToolbarItemOrientation;
}

export function ToolbarItem<T extends ToolbarItemOptions>({
  initialSelectedItem,
  items,
  onClick,
  orientation,
  type,
}: ToolbarItemProps<T>) {
  const [selectedItem, setSelectedItem] = useState(initialSelectedItem ?? items[0]);
  const [showMenu, setShowMenu] = useState<boolean>();
  const itemContainerRef = useRef<HTMLDivElement | null>(null);

  function handleItemClick(item: T) {
    type === 'mode-select' && setSelectedItem(item);
    onClick?.(item);
  }

  useEffect(() => {
    initialSelectedItem && setSelectedItem(initialSelectedItem);
  }, [initialSelectedItem]);

  useEffect(() => {
    items.length === 1 && setSelectedItem(items[0]);
  }, [items]);

  // When using a touch device, hide ToolbarItemMenu if clicking somewhere outside of it
  useEffect(() => {
    function handlePointerDown({ target }: Event) {
      if (showMenu && (!target || !itemContainerRef.current?.contains(target as HTMLDivElement))) {
        setShowMenu(false);
      }
    }

    document.addEventListener('pointerdown', handlePointerDown);

    return () => document.removeEventListener('pointerdown', handlePointerDown);
  }, [showMenu]);

  const menuItemComponents: JSX.Element[] | undefined = useMemo(() => {
    if (items.length > 1) {
      return items.reduce<JSX.Element[]>((accum, item, index) => {
        if ((type === 'action-select' && index > 0) || type !== 'action-select') {
          accum.push(
            <ItemContainer
              key={item.uuid}
              onItemClick={handleItemClick as (item: ToolbarItemOptions) => void}
              onPointerUp={() => {
                setShowMenu(false);
              }}
              item={item}
              type={type}
            />,
          );
        }

        return accum;
      }, []);
    }
    return undefined;
  }, [items, setShowMenu, handleItemClick, type]);

  return selectedItem ? (
    <ItemContainer
      onPointerDown={({ pointerType }) => {
        if (pointerType !== 'mouse' && type !== 'button') {
          setShowMenu(true);
        }
      }}
      onPointerEnter={({ pointerType }) => {
        if (pointerType === 'mouse' && type !== 'button') {
          setShowMenu(true);
        }
      }}
      onPointerLeave={({ pointerType }) => {
        if (pointerType === 'mouse' && type !== 'button') {
          setShowMenu(false);
        }
      }}
      onPointerUp={() => {
        type === 'button' && handleItemClick(selectedItem);
      }}
      item={{ ...selectedItem, text: undefined }} // don't display text for first level toolbar item
      type={type}
      disableSelectedStyle
      ref={itemContainerRef}
    >
      {type !== 'button' && menuItemComponents && (
        <Fragment>
          <CornerAdornment>{CornerTriangleSvg}</CornerAdornment>
          <ToolbarItemMenu isOpen={showMenu} orientation={orientation}>
            {menuItemComponents}
          </ToolbarItemMenu>
        </Fragment>
      )}
    </ItemContainer>
  ) : null;
}
