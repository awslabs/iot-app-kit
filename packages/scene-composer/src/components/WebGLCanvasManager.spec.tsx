import React from 'react';
import { Canvas, useThree } from '@react-three/fiber';
import { BoxGeometry, Mesh, MeshBasicMaterial, Group } from 'three';
import renderer from 'react-test-renderer';

import { useStore } from '../store';
import { setFeatureConfig } from '../common/GlobalSettings';

import { WebGLCanvasManager } from './WebGLCanvasManager';

import Mock = jest.Mock;

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useEffect: jest.fn(),
}));

// Mock other internal components
jest.mock('./three-fiber/EntityGroup', () => 'EntityGroup');
jest.mock('./three-fiber/Environment', () => {
  const originalModule = jest.requireActual('./three-fiber/Environment');
  return {
    ...originalModule,
    Environment: 'Environment',
  };
});
jest.mock('./three-fiber/StatsWindow', () => {
  return {
    StatsWindow: 'StatsWindow',
  };
});
jest.mock('./three-fiber/EditorCamera', () => {
  return {
    EditorMainCamera: 'EditorMainCamera',
  };
});
jest.mock('./three-fiber/EditorTransformControls', () => {
  return {
    EditorTransformControls: 'EditorTransformControls',
  };
});
jest.mock('./three-fiber/SceneInfoView', () => {
  return {
    SceneInfoView: 'SceneInfoView',
  };
});

// Mock Hooks
jest.mock('@react-three/fiber', () => {
  const originalModule = jest.requireActual('@react-three/fiber');
  return {
    ...originalModule,
    useThree: jest.fn(),
    useLoader: jest.fn(),
  };
});

class ResizeObserver {
  observe = jest.fn();
  unobserve = jest.fn();
  disconnect = jest.fn();
}

const Layout: React.FC = () => {
  return (
    <Canvas>
      <React.Suspense fallback={null}>
        <WebGLCanvasManager />
      </React.Suspense>
    </Canvas>
  );
};

describe('WebGLCanvasManagerSnap', () => {
  window.ResizeObserver = ResizeObserver;
  const body = document.createElement('body');
  const div = document.createElement('div');
  body.appendChild(div);

  const mockThreeStates = {
    gl: {
      domElement: div,
    },
    size: {
      width: 1920,
      height: 1080,
    },
  };

  const sceneDocument = {
    rootNodeRefs: ['testNode'],
    nodeMap: {
      testNode: { childRefs: ['childNode'] },
      childNode: { parentRef: 'testNode', childRefs: ['grandchildNode'] },
      grandchildNode: { parentRef: 'childNode', childRefs: [] },
    },
  };

  const group = new Group();
  const geometry = new BoxGeometry();
  geometry.uuid = 'Geometry';
  const material = new MeshBasicMaterial();
  material.uuid = 'Material';
  const mesh = new Mesh(geometry, material);
  group.name = 'Test';
  group.uuid = 'Forced';
  mesh.uuid = 'Forced2';
  group.add(mesh);

  const baseState: any = {
    getSceneNodeByRef: jest.fn(),
    getSceneProperty: jest.fn(),
    isEditing: jest.fn(),
    document: sceneDocument,
    getObject3DBySceneNodeRef: jest.fn().mockReturnValue(group),
  };

  beforeEach(() => {
    jest.clearAllMocks();

    const mockUseThree = useThree as Mock;
    mockUseThree.mockReturnValue(mockThreeStates);

    setFeatureConfig({});
  });

  it('should render editing correctly without immersive view feature', () => {
    useStore('default').setState(baseState);
    baseState.isEditing.mockReturnValue(true);
    baseState.getSceneNodeByRef.mockReturnValue('childNode');

    const container = renderer.create(<Layout />);
    expect(container).toMatchSnapshot();
  });

  it('should render viewing correctly without immersive view feature', () => {
    useStore('default').setState(baseState);
    baseState.isEditing.mockReturnValue(false);
    baseState.getSceneNodeByRef.mockReturnValue('childNode');

    const container = renderer.create(<Layout />);
    expect(container).toMatchSnapshot();
  });
});
