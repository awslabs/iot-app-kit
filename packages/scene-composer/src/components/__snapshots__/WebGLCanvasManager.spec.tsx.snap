// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebGLCanvasManagerSnap should render editing correctly without immersive view feature 1`] = `
<div
  style={
    {
      "height": "100%",
      "overflow": "hidden",
      "pointerEvents": "auto",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <div
    style={
      {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <canvas
      style={
        {
          "display": "block",
        }
      }
    />
  </div>
</div>
`;

exports[`WebGLCanvasManagerSnap should render viewing correctly without immersive view feature 1`] = `
<div
  style={
    {
      "height": "100%",
      "overflow": "hidden",
      "pointerEvents": "auto",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <div
    style={
      {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <canvas
      style={
        {
          "display": "block",
        }
      }
    />
  </div>
</div>
`;
